lib/uri/common.rb

```diff
@@ -62,22 +64,38 @@ def make_components_hash(klass, array_hash)
 
   include REGEXP
 
-  @@schemes = {}
+  SCHEME_LIST_MUTEX = Mutex.new
+  private_constant :SCHEME_LIST_MUTEX
+
   # Returns a Hash of the defined schemes.
+  # The list is lazily calculated.
   def self.scheme_list
-    @@schemes
+    return const_get(:SCHEMES) if defined?(SCHEMES)
+
+    SCHEME_LIST_MUTEX.synchronize do
+      const_set(:SCHEMES, ObjectSpace.
+        each_object(Class).
+        select { |klass| klass < URI::Generic }.
```

Comment:
This seems very expensive, iterate the whole heap when previously it was just registered directly.
Could the `Class#inherited` hook be used instead at least?
Or maybe use `Class#descendants` (checking if available with `respond_to?`) from https://bugs.ruby-lang.org/issues/14394?

In https://github.com/ruby/uri/pull/15 I think the way was clearly to use instance variables for this, and that has been accepted in https://bugs.ruby-lang.org/issues/17592. And the main Ractor could simply write to that instance variable. That has not been implemented yet though (cc @ko1).

Another strategy might be to haves `SCHEMES = {}`, and then once the first URI is created, `Ractor.make_shareable(SCHEMES) unless SCHEMES.frozen?`. That would prevent adding new schemes after the first URI is created, but maybe it's good enough, and it feels like it fits well with the Ractor model.

Adding support for Ractor by making everything slower/more hacky is IMHO absolutely not acceptable.